### ISSUE TITLE
The sclerp algorithm does not converge

### ISSUE THREAD

**Issue by wsp666** [Achllle/dual_quaternions#11]
We find that the interpolation algorithm fails to converge when two double quaternions are close, leading to a very outrageous result.
As shown below, the terminal feedback position T is
```
array([[-1.50605828e-01,  9.78038623e-01,  1.44077536e-01,
        -1.28873894e+03],
       [ 1.31112452e-01,  1.64213295e-01, -9.77672501e-01,
        -3.56061344e+02],
       [-9.79860913e-01, -1.28352817e-01, -1.52964521e-01,
        -2.03949883e+02],
       [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
         1.00000000e+00]])
```
here is the simplified code.
```python
from scipy.spatial.transform import Rotation
import numpy as np
from dual_quaternions import DualQuaternion

T1 = np.array([[-1.20911060e-02, -9.99300112e-01,  3.53990259e-02,
        -4.36968382e+00],
       [ 9.93674343e-01, -1.59607498e-02, -1.11160043e-01,
         1.54711835e+01],
       [ 1.11647238e-01,  3.38310559e-02,  9.93171865e-01,
         9.34843861e-01],
       [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
         1.00000000e+00]])

T2 = np.array([[-3.37726903e-02, -9.99344514e-01,  1.30364371e-02,
        -4.47583692e+00],
       [ 9.92807002e-01, -3.50451704e-02, -1.14481845e-01,
         1.54705164e+01],
       [ 1.14863668e-01,  9.07630609e-03,  9.93339800e-01,
         9.37977190e-01],
       [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
         1.00000000e+00]])
# Rotation matrices
R1 = T1[:3, :3]
R2 = T2[:3, :3]

# euler angles
euler1 = Rotation.from_matrix(R1).as_euler('zyx')
euler2 = Rotation.from_matrix(R2).as_euler('zyx')

print("Euler angles for T1: ", euler1)
print("Euler angles for T2: ", euler2)

dq1 = DualQuaternion.from_homogeneous_matrix(T1)
dq2 = DualQuaternion.from_homogeneous_matrix(T2)
delta = 0.5
# sclerp
dq = DualQuaternion.sclerp(dq1, dq2, delta)
T = dq.homogeneous_matrix()

# Linear interpolation is not satisfied
np.dot(np.linalg.inv(T1), T) - np.dot(T, np.linalg.inv(T2))
```

**Comment by RubensBenevides** [comment]
I'm facing exactly the same problem. I think there is something to do with the function 'transformation_matrix' called from pyquaternion, to convert from dual quaternion to matrice. In "Accessing matrix form" described here: "https://kieranwynn.github.io/pyquaternion/" they say:

"Both matrices and quaternions avoid the singularities and discontinuities involved with rotation in 3 dimensions by adding extra dimensions. This has the effect that different values could represent the same rotation, for example, quaternion q and -q represent the same rotation. It is therefore possible that, when converting a rotation sequence, the output may jump between different but equivalent forms. This could cause problems where subsequent operations such as differentiation are done on this data. Programmers should be aware of this issue."

So probably the map between matrices and quaternions is not working properly. I will check this elementwise on the conversion steps, if this is really the problem, I'll update my answer here.

**Comment by RubensBenevides** [comment]
After a lot of work examining every step inside the sclerp() function, I found the error. Just comment out these two lines and the error disappears:

```
if (start.q_r * stop.q_r).w < 0:
    start.q_r *= -1
```

I don't know exactly why the error stops occurring when we do this. I found out by pure testing. I will give more details so that someone who understands more about the mathematics of dual quaternions can explain it.

The sclerp() function will call the inverse() and pow() functions. The inverse() is working correctly, I have checked every step in the pyquaternion library, it is in pow() that the translation explodes. This happens exactly on line 13:

`se = (self.q_d.vector - s0 * d/2 * np.cos(theta/2)) / np.sin(theta/2)`

When dividing by `np.sin(theta/2)` the value becomes very large, and since it will be used to calculate the dual part of the dual quaternion, the translation value comes out wrong. I have no idea why this does not happen when we do not check the scalar part of the quaternion encoding the rotation `q_r.w`

All I know is that I tested the code in about 14 poses and it worked. The poses were very close in rotation and translation. Before removing those two lines, interpolation only worked on 4 of the 12 poses, now it works on all of them.

Translated with DeepL.com (free version)

**Comment by Achllle** [comment]
Thanks for digging in. I will take a look at this the coming weekend. I remember trying to fix this while working on #10 but getting stuck on something. Hopefully this moves it along! Otherwise PR welcome!

**Comment by RubensBenevides** [comment]
Where translation does not explode, SrLERP works perfectly like SLERP

**Comment by RubensBenevides** [comment]
The strange space of dual quaternions...
![I give up from qual quaternions](https://github.com/user-attachments/assets/491b6a49-5bc0-4c9b-9947-8ef97f495de5)

**Comment by RubensBenevides** [comment]
Man, there's something really, really strange going on in this code. I discovered that if I put the exp(cls, q) function after the sclerp() function, the interpolation result is different. What on earth could this influence in the code? The exp(cls, q) function isn't even used. Its original position is between pow() and sclerp(), I simply moved it down there, between lerp() and save(), and the interpolation result changed. I really freaked out about that. Then I realized that it was because I started making some changes to the nlerp function, changed its name to lerp and implemented the Dual Quaternion Linear Blending method (only for two quaternions, which is simply a linear interpolation with renormalization). Somehow, messing with the nlerp function, changes the result of sclerp. I'm really confused here, did changing the name of nlerp to lerp affect anything in pyquaternion?

**Comment by RubensBenevides** [comment]
Finally! In the pow function, change: 

`theta = 2*np.arccos(self.q_r.w)` 
to
`theta = dq_ratio.q_r.angle`

The pyquaternion function 'angle' will call atan2, which takes the quadrant into account when calculating the tangent arc, so somehow this makes a difference in the strange space of dual quaternions. Believe me, I've tested 64 combinations of signs between the elements of the dual quaternions being interpolated (start and stop in the sclerp). None worked, I even implemented a linear interpolation, because I'm only interested in the shortest path property, not the constant interval. To do this, I had to implement a normalization function for the dual quaternions, because the one in the library doesn't work. Both functions, my LERP and ScLERP, gave the same results, even when I interchanged the signs of the dual quaternions. 

Here is my pow() function:

```
import pyquaternion as quat
import dual_quaternion as dq

# Function used to calculate the power of a dual quaternion
def power_dq(dq_ratio, exponent):
    # Recover theta using the angle fuction from pyquaternion
    theta = dq_ratio.q_r.angle
    # Take the translation
    t = np.array(dq_ratio.translation())
    # If theta is zero return a dual quaternion with only the dual part
    if theta < 0.00001:
        print("Theta is too close to zero")
        return dq.DualQuaternion.from_translation_vector(exponent*np.array(dq_ratio.translation()))
    else:
        # screw direction (vector)
        s0 = dq_ratio.q_r.vector / np.sin(theta/2) 
        # translation along screw axis (scalar)
        d = -2. * dq_ratio.q_d.w / np.sin(theta/2)
        se = (dq_ratio.q_d.vector - s0 * d/2 * np.cos(theta/2)) / np.sin(theta/2)
    # Recover the interpolated quaternion of the rotation
    q_r = quat.Quaternion(scalar = np.cos(exponent*theta/2), 
                          vector = np.sin(exponent*theta/2) * s0)
    # Recover the interpolated quaternion of the translation (dual part)
    q_d = quat.Quaternion(scalar = -exponent*d/2 * np.sin(exponent*theta/2),
                          vector = exponent*d/2 * np.cos(exponent*theta/2) * s0 + np.sin(exponent*theta/2) * se)
    # create the dual quaternion
    dual_quat = dq.DualQuaternion(q_r, q_d)
    return dual_quat
```

here is my sclerp

```
def sclerp_test(start, stop, t):
    # Divide both dual quaternion: stop/start (by left division)
    division = start.quaternion_conjugate() * stop
    # Calculate the interpolated dual quaternion
    interpoled = start * power_dq(division,t)
    # Return the interpolated dual quaternion
    return interpoled
```

here is my normalization

```
def normalize_dual_quaternion(dual_quat):
    # Multiply by its conjugate to get a dual number QQ* = [s + t]
    squared_norm = dual_quat*dual_quat.quaternion_conjugate()   
    s = squared_norm.q_r.w
    t = squared_norm.q_d.w
    # Compute the dual number [a,b] = [a = 1/np.sqrt(s), b = (-tu^3)/2]
    # https://stackoverflow.com/questions/23174899/properly-normalizing-a-dual-quaternion
    a = 1 / (s)**0.5
    b = -t*(a**3) / 2
    # Multiply this dual number by the blended dual quaternion using dual number 
    # multiplication: (a+b)*(c+d) = (a*c) + (a*d + b*c)
    quat_r_normalized = a*dual_quat.q_r
    quat_d_normalized = a*dual_quat.q_d + b*dual_quat.q_r
    # Convert to dual quaternion
    dual_quat_normalized = dq.DualQuaternion(quat_r_normalized, quat_d_normalized)
    return dual_quat_normalized
```
and here is the lerp. I used it to compare with sclerp

```
def dq_lerp(start, stop, t):
    # Sum as dual quaternion
    blended_dq = (1-t)*start + t*stop
    # Normalize the blended_dq
    blended_dq_normalized = normalize_dual_quaternion(blended_dq)
    return blended_dq_normalized
```


In the sclerp, instead of using the start.inverse(), switching to start.quaternion_conjugate() gives the same result, so I opted for this because I think it's faster.

**Comment by RubensBenevides** [comment]
Sorry, I was celebrating too early, there is still some unexpected behavior, but the magnitude of the error has decreased a lot, to the point where I didn't notice it after applying the poses to the point clouds of a certain circuit. Before, the error would explode to more than 2,000 meters. Then it became like that image I posted earlier, around 200 meters, and now the error is between 10 and 15 meters in translation. That's pretty crazy. 

The image below shows how the interpolation should behave, it should return something very close to the black line (SLERP + LINEAR). At least now, in addition to having decreased by another order of magnitude, the error only occurs in the translation, the rotations are ok.

![image](https://github.com/user-attachments/assets/7f598977-78fe-442d-a996-eae3715c59fd)

**Comment by RubensBenevides** [comment]
I fixed it! Now, for the third time, I guarantee that this time it will be the last. This last error was a conditioning error, the one that happens when you divide by the sin(theta/2). I simply changed
`if np.isclose(theta, 0):`
by
`if theta < 0.000001:`
in the pow() function
and this is the result:
![image](https://github.com/user-attachments/assets/449e9a13-2c35-4c06-8e95-df6c353d6780)

**Comment by RubensBenevides** [comment]
Well, it is better to use 
`np.isclose(theta, 0)`
I tested:
`np.isclose(theta, 0, rtol=1e-6)`
`np.isclose(theta, 0, rtol=1e-5)`
`np.isclose(theta, 0, rtol=1e-4)`
`np.isclose(theta, 0, rtol=1e-3)`
`np.isclose(theta, 0, atol=1e-2)`
and 
`np.isclose(theta, 0, atol=1e-6)`
`np.isclose(theta, 0, atol=1e-5)`
`np.isclose(theta, 0, atol=1e-4)`
`np.isclose(theta, 0, atol=1e-3)`
`np.isclose(theta, 0, atol=1e-2)`
Amazing how they all don't work, but if I use 
`if abs(theta) < 0.0001`:
the function works

**Comment by Achllle** [comment]
That's awesome investigation, @RubensBenevides! Would you like to make a PR for this so it's fixed for everyone? 
Capturing your tests into unit tests would also be very welcome. If you don't have time for this, just let me know!

**Comment by RubensBenevides** [comment]
Unfortunately I don't have the time, and I don't even know how to do the pull request. Sorry.

**Comment by RubensBenevides** [comment]
A friend helped me make the pull request, so I did. But I don't know if everything is ok, do I need to put some examples there of what it looked like before and after?

**Comment by Achllle** [comment]
No it's great! I was meaning to test this out. Typically in a PR, you can make it easier for the maintainer to merge it in by providing test cases that show that the changes in the PR fix the issue. That also makes sure that in the future, no regressions happen by accident through some other change. Let me know if this is something you'd like to try, otherwise I'm happy to build that given all the examples in this issue.

**Comment by Achllle** [comment]
Solved with #13
